File: Practicals/Practical4-6/life_parallel_template/C/life_parallel_template.c
/*
*
*  Conway Game of Life
*
*  COMPILE: mpicc -o life life_parallel.c
*  RUN: mpirun -np 4 ./life
* 
*/

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#define NI 200 /* array sizes */
#define NJ 200  
#define NSTEPS 500 /* number of time steps */

void main(int argc, char **argv)
{
  int i, j, n, im, ip, jm, jp, ni, nj, nsum, isum, isumloc;
  int uniSize, indx;
  int *old, *new;
  float x;
  FILE *outFile;

  int nsdir, ewdir, ndims, periods[2], reorder;
  int centre,north,south,east,west;
  int coords[2], corner[2], dim_size[2];
  int iproc_nw, iproc_ne, iproc_sw, iproc_se;
  MPI_Comm comm_cart;
  MPI_Datatype row_hdle, col_hdle;
  MPI_Request req[16];


  MPI_Init(&argc,&argv);


  MPI_Comm_size(MPI_COMM_WORLD,&uniSize);

  if (uniSize != 4) {
    printf(" This code must be run with 4 processes\n");
    MPI_Abort(MPI_COMM_WORLD,10);
  }

  /* allocate arrays */
  ni = NI/2 + 2; /* add 2 for left and right ghost cells */
  nj = NJ/2 + 2;



  old = malloc(ni*nj*sizeof(int));
  new = malloc(ni*nj*sizeof(int));


//TODO: Create virtual topology
  nsdir = 0;
  ewdir = 1;

  ndims = ...
  dim_size[0]=... dim_size[1]=...
  periods[0] = ...
  periods[1] = ...
  reorder = ...
  MPI_Cart_create(...);
  MPI_Comm_rank(...);
  MPI_Cart_shift(...);
  MPI_Cart_shift(...);


  MPI_Cart_coords(...);

  /*  initialize elements of old to 0 or 1    *
   *  Set the seed like this so that it can   *
   *  directly compared with the serial       *
   *  version                                 */
  for(i=1; i<ni-1; i++) {
    j = i + coords[0]*(nj-2);
    srand(1234+j);

    if (coords[1] == 1) {
       for (j=1; j<nj-1; j++) rand();
    }
    for(j=1; j<nj-1; j++) {
      indx = j + nj*i;
      x = rand()/((float)RAND_MAX + 1);
      if(x < 0.5) {
          old[indx] = 0;
      } else {
          old[indx] = 1;
      }
    }

    if (coords[0] == 0) {
       for (j=1; j<nj-1; j++) rand();
    }
  }



  isum = 0;
  for (i=1; i<ni-1; i++) {
    for (j=1; j<nj-1; j++) {
      indx = j + nj*i;
      isum = isum + old[indx];
    }
  }
  printf("Initial sum %d, %d\n",centre,isum);
  isumloc = isum;
  MPI_Reduce(&isumloc,&isum,1,MPI_INT,MPI_SUM,0,comm_cart);
  if (centre == 0) printf(" Init total %d\n",isum);

  printf(" Neighbours %d, %d, %d, %d, %d\n",centre,north,south,east,west);


  corner[0] = (coords[0]+1)%2; 
  corner[1] = (coords[1]+1)%2; 
  MPI_Cart_rank(comm_cart,corner,&iproc_ne);

/* In this case corners are all the smae rank */
  iproc_nw = iproc_ne;
  iproc_sw = iproc_ne;
  iproc_se = iproc_ne;


//TODO: Create derived type for single row and column 

  MPI_Type_vector(...);
  MPI_Type_commit(...);

  MPI_Type_vector(...);
  MPI_Type_commit(...);



  /*  time steps */
  for(n=0; n<NSTEPS; n++) {



/*TODO: send and receive north and south rows */
       MPI_Isend(...);          //  row=1, col=1
       MPI_Irecv(...);             //  row=0, col=1
       MPI_Isend(...);   //  row=ni-2, col=1
       MPI_Irecv(...);   //  row=ni-1, col=1

/*TODO: send and receive west and east columns */
       MPI_Isend(...);           //  row=1, col=1 
       MPI_Irecv(...);             //  row=1, col=0
       MPI_Isend(...);         //  row=1, col=nj-2
       MPI_Irecv(...);         //  row=1, col=nj-1

/*TODO: send and receive corners */
       MPI_Isend(...);          // row=1, col=1
       MPI_Irecv(...);             // row=0, col=0
       MPI_Isend(...);       // row=1, col=nj-2
       MPI_Irecv(...);         // row=0, col=nj-1
       MPI_Isend(...);  // row=ni-2, col=1
       MPI_Irecv(...);    // row=ni-1, col=0
       MPI_Isend(...); // row=ni-2, col=nj-2
       MPI_Irecv(...); // row=ni-1, col=nj-1 


       MPI_Waitall(...);





    for(i=1; i<ni-1; i++) {
      for(j=1; j<nj-1; j++) {

        im = i-1;
        ip = i+1;
        jm = j-1;
        jp = j+1;

        nsum =  old[im*nj+jp] + old[i*nj+jp] + old[ip*nj+jp]
              + old[im*nj+ j] + old[ip*nj+j]
              + old[im*nj+jm] + old[i*nj+jm] + old[ip*nj+jm];

        switch(nsum) {

          case 3:
            new[i*nj+j] = 1;
            break;

          case 2:
            new[i*nj+j] = old[i*nj+j];
            break;

          default:
            new[i*nj+j] = 0;
        }
      }
    }

    /* copy new state into old state */
    for(i=1; i<ni-1; i++) {
      for(j=1; j<nj-1; j++) {
        old[i*nj+j] = new[i*nj+j];
      }
    }
  }

  /*  Iterations are done; sum the number of live cells */
  isum = 0;
  for(i=1; i<ni-1; i++) {
    for(j=1; j<nj-1; j++) {
      isum = isum + new[i*nj+j];
    }
  }


  printf("\nNumber of live cells = %d\n", isum);

  isumloc = isum;
  MPI_Reduce(&isumloc,&isum,1,MPI_INT,MPI_SUM,0,comm_cart);
  if (centre == 0) printf(" Final total %d\n",isum);

  MPI_Type_free(&row_hdle);
  MPI_Type_free(&col_hdle);

  MPI_Comm_free(&comm_cart);
  MPI_Finalize();
}

File: Practicals/Practical4-6/life_parallel_template/Fortran/life_parallel_template.f90
!------------------------------------
!  Conway Game of Life

! 4 processors, domain decomposition
! in both i and j directions, virtual
! topology

!------------------------------------
program life
  use mpi_f08
  implicit none

  integer (kind=4), parameter :: ni=200, nj=200, nsteps=500, rank=1
  integer :: ndims,dim_size(2),coords(2),corner(2),uniSize,nsdir,ewdir
  integer :: centre,north,south,east,west,iproc_nw,iproc_ne,iproc_sw,iproc_se
  integer :: seed_size
  integer (kind=4) :: isum,isumloc,ninom,njnom
  integer (kind=4) :: i,j,ip,jp,im,jm,n,nsum
  integer, allocatable :: rnd_seed(:), row(:)
  integer (kind=4), allocatable, target, dimension(:,:) :: old, new
  real, allocatable, dimension(:) :: arand
  logical :: reorder, periods(2)
  type(MPI_Comm) :: comm_cart
  type(MPI_Datatype) :: row_hdle, col_hdle
  type(MPI_Request) :: req(16)


  call MPI_Init()
  call MPI_Comm_size(MPI_COMM_WORLD,uniSize)
  if (uniSize .NE. 4) then
    write(6,*) ' Must have 4 MPI processes exactly '
    call MPI_Abort(MPI_COMM_WORLD,10)
  endif


 
  ! nominal number of points per proc. (without ghost cells,
  ! assume numbers divide evenly); dim_size(1) and dim_size(2)
  ! are the numbers of procs in the i and j directions.


  dim_size(1) = 2
  dim_size(2) = 2

  ninom = ni/dim_size(1)
  njnom = nj/dim_size(2)


!TODO: Create virtual topology
  ! presently set up for only 4 procs
  ! periodic in all directions

  ! coordinate directions for north, south, east, and west
  nsdir = 0
  ewdir = 1

  ndims = ...
  periods(1) = ....
  periods(2) = ...
  reorder = ...
  call mpi_cart_create(...)
  call MPI_Comm_rank(...)
  call mpi_cart_shift(...)
  call mpi_cart_shift(...)



!TODO: Define processors on corners
  ! Virtual topology has been used to define adjacent processors
  ! to the north, south, east, and west; still have to define
  ! processors on corners

  call MPI_cart_coords(...)
! North-west
  corner(1) = modulo(coords(1)+1,2); corner(2) = modulo(coords(2)-1,2); 
  call mpi_cart_rank(...)

! In this case corners are all the same rank
  iproc_ne = iproc_nw
  iproc_sw = iproc_nw
  iproc_se = iproc_nw


  if (centre .EQ. 0) write(6,*) ' Neighbours ',north,south,east,west  
  ! allocate arrays

  allocate(old(0:ninom+1,0:njnom+1),new(0:ninom+1,0:njnom+1))
  allocate(arand(ni),row(njnom))

  old = 0; new = 0;


  ! Initialize elements of old to 0 or 1.  We're doing some
  ! sleight of hand here to make sure we initialize to the
  ! same values as in the serial case. The random_number
  ! function is called for every i and j, even if they are
  ! not on the current processor, to get the same random
  ! distribution as the serial case, but they are only used
  ! if this i and j reside on the current procesor.


  call random_seed(SIZE=seed_size)
  allocate(rnd_seed(seed_size))
  call random_seed(GET=rnd_seed)

  rnd_seed(1) = rnd_seed(1) - coords(2)*ninom
  do j = 1, njnom
    rnd_seed(1) = rnd_seed(1) - 1
    call random_seed(PUT=rnd_seed)
    call random_number(arand)
    do i = 1,ninom 
      old(i,j) = nint(arand(coords(1)*ninom+i))
    end do
  enddo


  isum = sum(old(1:ninom,1:njnom))
  write(6,*) ' Initial sum ',centre,isum
  isumloc = isum
  isum = 0
  call mpi_reduce(isumloc,isum,1,MPI_INTEGER4,MPI_SUM,0,comm_cart)
  if (centre .EQ. 0) write(6,*) ' Init total ',isum


! TODO: Create derived type for single row of array.
  ! There are nj "blocks," each containing 1 element,
  ! with a stride of nitot between the blocks

  call mpi_type_vector( ... )
  call mpi_type_commit(...)

  call mpi_type_vector(...)
  call mpi_type_commit(...)


  !  iterate

  time_iteration: do n = 1, nsteps

     ! transfer data to ghost cells



!TODO: send and receive top and bottom rows 
        call mpi_isend(...)          ! row=1, col=1
        call mpi_irecv(...)          ! row=0, col=1
        call mpi_isend(...)      ! row=ninom, col=1
        call mpi_irecv(...)    ! row=ninom+1, col=1

!TODO: send and receive left and right columns 
        call mpi_isend(...)           ! row=1,col=1
        call mpi_irecv(...)           ! row=0, col=1
        call mpi_isend(...)       ! row=1, col=njnom
        call mpi_irecv(...)     ! row=1, col=njnom+1

!TODO: send and receive corners 
        call mpi_isend(...)
        call mpi_irecv(...)
        call mpi_isend(...)
        call mpi_irecv(...)
        call mpi_isend(...)
        call mpi_irecv(...)
        call mpi_isend(...)
        call mpi_irecv(...)

!TODO: make sure all non-blocking messages have arrived
        call mpi_waitall(16,req,MPI_STATUSES_IGNORE)


     ! update states of cells

     do j = 1, njnom
        do i = 1, ninom

           ip = i + 1
           im = i - 1
           jp = j + 1
           jm = j - 1

           nsum =  old(im,jp) + old(i,jp) + old(ip,jp) &
                 + old(im,j )             + old(ip,j ) &
                 + old(im,jm) + old(i,jm) + old(ip,jm)

           select case (nsum)
           case (3)
              new(i,j) = 1
           case (2)
              new(i,j) = old(i,j)
           case default
              new(i,j) = 0
           end select

        enddo
     enddo

     ! copy new state into old state
     old = new


  enddo time_iteration


  ! Iterations are done; sum the number of live cells
  
  isum = sum(new(1:ninom,1:njnom))
  
  ! Print final number of live cells

  isumloc = isum
  isum = 0
  call mpi_reduce(isumloc,isum,1,MPI_INTEGER4,MPI_SUM,0,comm_cart)

  if(centre == 0) then
     write(*,"(/'Number of live cells = ', i6/)") isum
  endif

  call MPI_Type_free(row_hdle)
  call MPI_Type_free(col_hdle)

  deallocate(old, new, arand)


  call MPI_Comm_free(comm_cart)
  call mpi_finalize()

end program life
